shell/platform/embedder/tests/embedder_unittests.cc

```diff
@@ -2373,5 +2380,181 @@ TEST_F(EmbedderTest, VerifyB141980393) {
+/// Test that an engine can be deinitialized.
+///
+TEST_F(EmbedderTest, CaDeinitializeAnEngine) {
+  EmbedderConfigBuilder builder(GetEmbedderContext());
+  builder.SetSoftwareRendererConfig();
+  auto engine = builder.InitializeEngine();
+  ASSERT_TRUE(engine.is_valid());
+  ASSERT_EQ(FlutterEngineRunInitialized(engine.get()), kSuccess);
+  // Cannot re-run an already running engine.
+  ASSERT_EQ(FlutterEngineRunInitialized(engine.get()), kInvalidArguments);
+  ASSERT_EQ(FlutterEngineDeinitialize(engine.get()), kSuccess);
+  // It is ok to deinitialize an engine multiple times.
+  ASSERT_EQ(FlutterEngineDeinitialize(engine.get()), kSuccess);
+
+  // Sending events to a deinitalized engine fails.
+  FlutterWindowMetricsEvent event = {};
+  event.struct_size = sizeof(event);
+  event.width = 800;
+  event.height = 600;
+  ASSERT_EQ(FlutterEngineSendWindowMetricsEvent(engine.get(), &event),
+            kInvalidArguments);
+  engine.reset();
+}
+
+//------------------------------------------------------------------------------
+/// Asserts that embedders can provide a task runner for the render thread.
+///
+TEST_F(EmbedderTest, CanCreateEmbedderWithCustomRenderTaskRunner) {
+  std::mutex engine_mutex;
+  UniqueEngine engine;
+  fml::AutoResetWaitableEvent task_latch;
+  bool task_executed = false;
+  EmbedderTestTaskRunner render_task_runner(
+      CreateNewThread("custom_render_thread"), [&](FlutterTask task) {
+        std::scoped_lock engine_lock(engine_mutex);
+        if (engine.is_valid()) {
+          ASSERT_EQ(FlutterEngineRunTask(engine.get(), &task), kSuccess);
+          task_executed = true;
+          task_latch.Signal();
+        }
+      });
+  EmbedderConfigBuilder builder(GetEmbedderContext());
+  builder.SetDartEntrypoint("can_render_scene_without_custom_compositor");
+  builder.SetOpenGLRendererConfig(SkISize::Make(800, 600));
+  builder.SetRenderTaskRunner(
+      &render_task_runner.GetFlutterTaskRunnerDescription());
+
+  {
+    std::scoped_lock lock(engine_mutex);
+    engine = builder.InitializeEngine();
+  }
+
+  ASSERT_EQ(FlutterEngineRunInitialized(engine.get()), kSuccess);
+
+  ASSERT_TRUE(engine.is_valid());
+
+  FlutterWindowMetricsEvent event = {};
+  event.struct_size = sizeof(event);
+  event.width = 800;
+  event.height = 600;
+  ASSERT_EQ(FlutterEngineSendWindowMetricsEvent(engine.get(), &event),
+            kSuccess);
+  task_latch.Wait();
+  ASSERT_TRUE(task_executed);
+  ASSERT_EQ(FlutterEngineDeinitialize(engine.get()), kSuccess);
+
+  {
+    std::scoped_lock engine_lock(engine_mutex);
+    engine.reset();
+  }
+}
+
+//------------------------------------------------------------------------------
+/// Asserts that the render task runner can be the same as the platform task
+/// runner.
+///
+TEST_F(EmbedderTest,
+       CanCreateEmbedderWithCustomRenderTaskRunnerTheSameAsPlatformTaskRunner) {
+  // We need to create a new thread to act as the platform thread because we
```

Comment:
nit: avoid "we" and remove extra line below?

Reply:
Done.